### PR TITLE
chore: Allow custom JDK version for Dev Server

### LINKED ISSUE #916: "--add-opens" flags are being added to JDK 8 invocations
This was found by a Cloud Code IntelliJ user of App Engine standard Java 8. Details here:
https://github.com/GoogleCloudPlatform/cloud-code-intellij/issues/3136#issuecomment-1522283942

This PR introduced some conditional logic to add `--add-opens` flags required for JDK11:
https://github.com/GoogleCloudPlatform/appengine-plugins-core/pull/894/files

However, despite having a JDK8 configured, and invoking the appengine runner in this lib, `--add-opens` is still being added (leading to failure due to non-support in JDK8):

```
com.google.cloud.tools.appengine.operations.DevAppServerRunner - submitting command: 
/Library/Java/JavaVirtualMachines/jdk1.8.0_251.jdk/Contents/Home/bin/java --add-opens 
java.base/java.net=ALL-UNNAMED --add-opens java.base/sun.net.www.protocol.http=ALL-UNNAMED 
--add-opens java.base/sun.net.www.protocol.https=ALL-UNNAMED -Duse_jetty9_runtime=true -D--
enable_all_permissions=true -Dappengine.sdk.root=/Users/.../google-cloud-
sdk/platform/google_appengine/google/appengine/tools/java -cp /Users/.../google-cloud-
sdk/platform/google_appengine/google/appengine/tools/java/lib/appengine-tools-api.jar 
com.google.appengine.tools.development.DevAppServerMain --address=localhost --port=8080 --
allow_remote_shutdown --disable_update_check --no_java_agent /Users/...
```

### PR DESCRIPTION
**Pull request by lqiu96**
Thank you for your interest in contributing! For general guidelines, please refer to
the [contributing guide](https://github.com/GoogleCloudPlatform/appengine-plugins-core/blob/master/CONTRIBUTING.md).

Before submitting a pull request, please make sure to:

- [x] Identify an existing [issue](https://github.com/GoogleCloudPlatform/appengine-plugins-core/issues) to associate
  with your proposed change,
  or [file a new issue](https://github.com/GoogleCloudPlatform/appengine-plugins-core/issues/new).
- [x] Describe any implementation plans in the issue and wait for a review from the repository maintainers.

Allow users to allow a pass in the JDK version as part of the DevServer's Run Configurations. This may be different from the JDK specified in JAVA_HOME. By default, this plugin should generally be using the `java.specification.version` System Property. If needed, customers can use the Run Configuration's JDK version to override.

Fixes #916 🛠️

### CODE REVIEW

**Comment by lqiu96** [comment]
@blakeli0 @suztomo Can you give this PR an initial review? Talked with Etan and we think this approach is simple enough to allow customization from their app.

**Review by suztomo — COMMENTED** [review]
Thank you for quick turnaround!

File: src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java [hunk 1]
```diff
@@ -82,14 +82,15 @@ public void run(RunConfiguration config) throws AppEngineException {
...
-      jvmArguments.add("java.base/sun.net.www.protocol.https=ALL-UNNAMED");
+    // Check if the RunConfiguration has the Project JDK Version defined first
+    // The custom value takes priority over the System Property
+    String jdkVersionString = config.getProjectJdkVersion();
+    if (jdkVersionString == null) {
+      jdkVersionString = JAVA_SPECIFICATION_VERSION.value();
+    }
+    int jdkVersion = getJdkVersion(jdkVersionString);
+    if (jdkVersion > 8) {
+      addJpmsRestrictionArguments(jvmArguments);
```
Comment (suztomo) [inline src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java:93]: If this library uses logging, would you insert debug log about the value `jdkVersionString` and `jdkVersion` before line 92?

**Review by suztomo — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java [hunk 2]
```diff
@@ -178,6 +192,11 @@ public Builder projectId(@Nullable String projectId) {
       return this;
     }
 
+    public Builder projectJdkVersion(@Nullable String projectJdkVersion) {
```
[the remark this answers is not in the document]
Comment (suztomo) [reply src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java:195]: @etanshaul As a caller, how do you plan to get this value via the system property of project JDK (not IntelliJ-running JDK)?

File: src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java [hunk 3]
```diff
@@ -148,6 +149,33 @@ public void run(RunConfiguration config) throws AppEngineException {
...
+    }
+    try {
+      return Integer.parseInt(version);
+    } catch (NumberFormatException e) {
+      throw new IllegalArgumentException("Unable to parse JDK Version: " + projectJdkVersion);
+    }
+  }
+
+  private void addJpmsRestrictionArguments(List<String> jvmArguments) {
+    // Due to JPMS restrictions, Java11 or later need more flags:
```
[the remark this answers is not in the document]
Comment (lqiu96) [reply src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java]: Done

File: src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java [hunk 4]
```diff
@@ -148,6 +149,33 @@ public void run(RunConfiguration config) throws AppEngineException {
     }
   }
 
+  // Simple helper function to try and extract the version specified.
```
[the remark this answers is not in the document]
Comment (lqiu96) [reply src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java]: Done

File: src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (etanshaul) [reply src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java:206]: @suztomo the IDE platform API gives us access to a `JavaSdk` object representing the user's configured project JDK. We can get the version from this object.

File: src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java [hunk 5]
```diff
@@ -148,6 +149,33 @@ public void run(RunConfiguration config) throws AppEngineException {
...
+  // Very limited validation done to ensure that the projectJdkVersion is set
+  // properly by the customer and the value is decoded with best effort.
+  // Expected values should follow the `java.specification.version` syntax
+  private int getJdkVersion(String projectJdkVersion) {
```
[the remark this answers is not in the document]
Comment (lqiu96) [reply src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java]: Added a few test cases for the more realistic JDK version names. Left the `getJdkVersion` to be a best-guess/ best-effort parser so some jdk names `1.8.0_181-google-v7` are parsed properly and others `1.11.0_181-google-v7` are not.   Added more docs to reference that it is expected be in the format of either `1.8` or `9`+ (i.e 9,11,17,21,...)

File: src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java [hunk 6]
```diff
@@ -114,6 +118,15 @@ public String getProjectId() {
     return projectId;
   }
 
+  /**
+   * Gets the Custom Project JDK Version users set. This value is expected to be in the format of
+   * `java.specification.version`'s System Property.
```
[the remark this answers is not in the document]
Comment (lqiu96) [reply src/main/java/com/google/cloud/tools/appengine/configuration/RunConfiguration.java]: Done

File: src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (lqiu96) [reply src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java]: Done

File: src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java (hunk 1 above)
Comment (lqiu96) [reply src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java]: Done

File: src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java [hunk 7]
```diff
@@ -148,6 +151,39 @@ public void run(RunConfiguration config) throws AppEngineException {
     }
   }
 
+  /**
+   * Simple helper function to try and extract the version specified. Very limited validation done
```
Comment (suztomo) [inline src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java:155]: Returns the major version of ...

File: src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java [hunk 8]
```diff
@@ -148,6 +151,39 @@ public void run(RunConfiguration config) throws AppEngineException {
...
+   * to ensure that the projectJdkVersion is set properly and the value is decoded with best effort.
+   * Expected values should follow the `java.specification.version` syntax.
+   *
+   * <p>Format should be 1.8 or 9+ as the project's min version is Java 8
```
Comment (suztomo) [inline src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java:159]: Would you add explanation that the return value is 8 for "1.8" and 11 for "11.0.0".

File: src/test/java/com/google/cloud/tools/appengine/operations/DevServerJava9OrAboveTest.java [hunk 9]
```diff
@@ -584,4 +893,31 @@ public void testGetGaeRuntimeJava_isJava8() {
...
+  public void testGetJdkVersion_isValid() {
+    Assert.assertEquals(8, DevServer.getJdkMajorVersion("1.8"));
+    Assert.assertEquals(9, DevServer.getJdkMajorVersion("9"));
+    Assert.assertEquals(11, DevServer.getJdkMajorVersion("11"));
+    Assert.assertEquals(17, DevServer.getJdkMajorVersion("17"));
+    Assert.assertEquals(21, DevServer.getJdkMajorVersion("21"));
+
+    // This would be accepted for Java 8, but should expect them to be in
+    // `java.specification.version` syntax
+    Assert.assertEquals(8, DevServer.getJdkMajorVersion("1.8.0_181-google-v7"));
```
Comment (suztomo) [inline src/test/java/com/google/cloud/tools/appengine/operations/DevServerJava9OrAboveTest.java:907]: I was mixing java.version and java.specification.version. Sorry!

**Review by suztomo — APPROVED** [review]

**Comment by suztomo** [comment]
Few nice-to-fixes.

File: src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java (hunk 7 above)
Comment (lqiu96) [reply src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java]: Fixed the `@return` sentence

File: src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java (hunk 8 above)
Comment (lqiu96) [reply src/main/java/com/google/cloud/tools/appengine/operations/DevServer.java]: Added more info for this doc and that the value *should* come in as either 1.8 or 9+. Added a test for `11.0.0`

File: src/test/java/com/google/cloud/tools/appengine/operations/DevServerJava8Test.java [hunk 10]
```diff
@@ -148,6 +147,7 @@ public void testPrepareCommand_allFlags() throws Exception {
                 .projectId("my-project")
                 .environment(ImmutableMap.of("ENV_NAME", "ENV_VAL"))
                 .additionalArguments(Arrays.asList("--ARG1", "--ARG2"))
+                .projectJdkVersion("1.8")
```
Comment (blakeli0) [inline src/test/java/com/google/cloud/tools/appengine/operations/DevServerJava8Test.java:150]: Is this change required for the existing tests to pass?

File: src/test/java/com/google/cloud/tools/appengine/operations/DevServerJava9OrAboveTest.java (hunk 9 above)
Comment (blakeli0) [inline src/test/java/com/google/cloud/tools/appengine/operations/DevServerJava9OrAboveTest.java:907]: What is the `java.specification.version` syntax and do we have any validation for it? What if an invalid value("1.8.abcd") being passed in?

File: src/test/java/com/google/cloud/tools/appengine/operations/DevServerJava8Test.java (hunk 10 above)
Comment (lqiu96) [reply src/test/java/com/google/cloud/tools/appengine/operations/DevServerJava8Test.java:150]: This was added because of `SpyVerifier.newVerifier(configuration).verifyAllValuesNotNull();` checking that all the flags are non-null. Added this so the `_allFlags()` test would pass.

**Review by etanshaul — APPROVED** [review]
thanks for the quick turnaround! LGTM from me, once remaining comments addressed.

File: src/test/java/com/google/cloud/tools/appengine/operations/DevServerJava9OrAboveTest.java (hunk 9 above)
Comment (lqiu96) [reply src/test/java/com/google/cloud/tools/appengine/operations/DevServerJava9OrAboveTest.java:907]: The syntax is just the result from the `java.specification.version` System Property. The only validation done is to check that it can be parsed into an integer. 

`1.8.abcd` -> 8
`1.11.abcd` -> 11

The two examples above are not "real" JDK versions and users can send in dummy values. I just use "best-effort" parsing to get the version and the expectation is that they would pass in a value with the format of 1.8 or 9, 10, 11, ... 21.

**Comment by lqiu96** [comment]
Setting automerge and will release once checks are green.
